Nondet-initialize enums to be equal to a constant of the same type

antlechner opened the pull request:
Previously we nondet-initialized enums just like any other type: by nondet-initializing each field. In the case of enums, there are two fields: `String name` and `int ordinal`. These two fields depend on each other, for example, for an enum
```public enum Color { RED, GREEN, BLUE }```
the enum `Color.BLUE` would have `name == "BLUE"` and `ordinal == 2`. Java generally only allows an enum object to be either null, or equal to one of the constants defined for its type.

This PR introduces this restriction in `java_object_factory`. The changed part of the generated GOTO program looks e.g. like this:
```
        struct Color *c;
        int enum_index_init;
        clinit_wrapper();
        IF !NONDET(_Bool) THEN GOTO 1
        c = null;
        GOTO 2
     1: enum_index_init = NONDET(int);
        ASSUME enum_index_init >= 0
        ASSUME enum_index_init < $VALUES->length
        c = (struct Color *)$VALUES->data[enum_index_init];
     2: INPUT("c", c);
```

A corresponding change in lazy methods v1 is also necessary as it has to mirror the behaviour of __CPROVER_start.

antlechner committed:
CI lazy methods: load clinit for all param types

We load static initializers for all types of parameters of the entry
function(s). This is necessary for the new nondet-initialization of
enums, and also consistent with the Java language specification, which
says that static initializers should be run on object initialization.

antlechner committed:
Add some documentation to java_object_factory

Add a high-level overview of the recursive algorithm in the file.

antlechner committed:
Always run clinit_wrapper before nondet object init

This is more consistent with the behaviour of the JDK (the Java language
specification says that static initializers should be run on object
initialization), and a necessary requirement to access enum constants.

antlechner committed:
Refactor logic for generating a nondet int

antlechner committed:
Nondet-init enums by assigning them to a constant

Rather than nondet-initializing each field (String name and int ordinal)
of an enum, we assign it to be equal to one of the constant values
stored in the $VALUES array of the corresponding type.

antlechner committed:
Add tests for nondet initialization of enums

The tests check that for a nondeterministic non-null enum, any of the
constants of its type can be chosen, and nothing other than these
constants.

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -756,6 +761,24 @@ void java_object_factoryt::gen_nondet_pointer_init(
+    const java_class_typet &class_type = to_java_class_type(subtype);
+    const irep_idt &class_name = class_type.get_name();
+    const irep_idt class_clinit = clinit_wrapper_name(class_name);
+    gen_method_call(assignments, expr, class_clinit);
```
:question: I think @smowton should see this - he spent a bunch of time thinking about when static inits should be run. I don't disagree with your conclusion BTW, but would like them to weigh in.

antlechner replied:
Yes please @smowton ! I raised exactly this question at a few stand-ups, and @peterschrammel said it's probably best to call static initialisers for all argument types, not just enums, as that's closer to what "real Java" does. No one could really think of an example where it would really make a difference for our generated tests, or indeed why the JVM calls static initialisers before constructor calls. @peterschrammel suggested it may have something to do with concurrency, so @cesaro might have an opinion on this too.

thomasspriggs replied:
The JVM should call static initialisers before constructors, because the constructor could reference a statically initialised value.

antlechner replied:
@thomasspriggs But then the rule might as well be "the JVM should call static initialisers before referencing statically initialised values"?

thomasspriggs replied:
@antlechner But then the JVM would have to keep track of which values are statically initialised. I suspect that just keeping track of which classes have had their static initialiser run is simpler.

smowton replied:
"Run clinit before introducing an instance of a given type" seems sensible to me. It is indeed impossible to observe a class' static fields without triggering a static init in any case, but the order they are executed will vary depending on whether you follow the clinit-on-construction rule or not.

smowton replied:
Beware this might reduce coverage though if the type's static fields were unused. We may want an option to break the JVM spec in this respect and only static init upon actual static field reference.

antlechner replied:
Good point about coverage.
@peterschrammel Should I leave this change as it is for now, and if it reduces coverage then we'll add an option for it? Or should I introduce the option right away, just in case? Or put everything behind a "if enum type"?

antlechner replied:
Also, if we did have an option, shouldn't it apply to calling `clinit_wrapper` before regular/deterministic constructor calls too?

peterschrammel replied:
> We may want an option to break the JVM spec in this respect and only static init upon actual static field reference.

We should try that to make it more a conscious choice to deviate.

antlechner replied:
@peterschrammel As part of this PR for nondet-initialisation only, or in a future PR for both nondet and det object initialisation?

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1599,3 +1608,24 @@ void gen_nondet_init(
+  const exprt &instance_expr,
+  const irep_idt &method_name)
+{
+  if(const auto func = symbol_table.lookup(method_name))
```
:bulb: You might prefer to make this a `PRECONDITION` as almost certainly an error if the method doesn't exist and silently failing is normally bad

antlechner replied:
Actually it will usually be the case that the method doesn't exist, as most classes don't have both a static initialiser and a `cproverNondetInitialize`, which are the values for `method_name` that this is called with. I renamed this function to `gen_method_call_if_present` to clarify this.

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1201,6 +1207,56 @@ void java_object_factoryt::gen_nondet_init(
+/// \param min_value_expr: Represents the minimum value for the integer.
+/// \param max_value_expr: Represents the maximum value for the integer.
+/// \return A symbol expression for the resulting integer.
+const symbol_exprt java_object_factoryt::gen_nondet_int_init(
```
:bulb: this is similar to `generate_nondet_int` in `nondet.cpp` in TG, it might be nice to de-duplicate the calls

antlechner replied:
Yes, I talked to @thomasspriggs about this and I'm planning to make this change in a follow-up PR, should have mentioned it here.

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1201,6 +1207,56 @@ void java_object_factoryt::gen_nondet_init(
+{
+  // Allocate a new symbol for the int
+  const symbolt &int_symbol = get_fresh_aux_symbol(
+    java_int_type(),
```
:no_entry_sign: Either use the type of the `min_value_expr` and `max_value_expr` or `PRECONDITION(min_value_expr.type() == java_int_type()); PRECONDITION(max_value_expr.type() == java_int_type())` otherwise I think your assumptions are not well formed.

antlechner replied:
Done (using the type of the `min_value_expr` here, and a precondition above making sure that `max_value_expr` has the same type).

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1201,6 +1207,56 @@ void java_object_factoryt::gen_nondet_init(
+  const auto &int_symbol_expr = int_symbol.symbol_expr();
+
+  // Nondet-initialize it
+  gen_nondet_init(
```
:bulb: Might be simpler to nondet it directly here rather than using this massive function tbh.

thk123 replied:
```C++
  assignments.add(
    code_assignt(int_symbol_expr, side_effect_expr_nondett(int_type)));
```

antlechner replied:
Hmm, not sure about this one. On the one hand, every function call to `gen_nondet_init` looks horrible because it takes so many parameters. On the other hand, it's nice to just rely on `gen_nondet_init` to generate _anything_ for you, and I think it clarifies its role in this file when it is used this way.
Also, for primitive types it also has a `add_source_location() = loc` operation in addition to what you wrote, so that would become more like four lines overall and not be much more readable than the current version.

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1220,45 +1276,18 @@ void java_object_factoryt::allocate_nondet_length_array(
-
-  // Initialize array with some undetermined length:
-  gen_nondet_init(
+  const auto &length_sym_expr = gen_nondet_int_init(
```
:green_heart:

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -146,7 +146,8 @@ class java_object_factoryt
     const exprt &min_length_expr,
-    const exprt &max_length_expr);
+    const exprt &max_length_expr,
+    bool include_max);
```
:no_entry_sign: ~Make `max_length_expr` `optionalt` then~ so this parameter is for inclusive range, use the word `inclusive_max`. 

:bulb: I think I'd prefer to handle this at the callsite (i.e. decrement the max to make it non-inclusive). 
:bulb: Alternatively, create an enum for the options:

```
enum class int_bonudst
{
  inclusive_bounds, 
  exclude_min,
  exclude_max,
  exclusive_bounds
}
```
(or as a bit flag if you'd prefer).

antlechner replied:
> I think I'd prefer to handle this at the callsite (i.e. decrement the max to make it non-inclusive).

Done, using a `minus_exprt` in the call, you might want to check if this is correct.

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -384,67 +385,95 @@ void java_object_factoryt::gen_pointer_target_init(
 {
-  PRECONDITION(expr.type().id()==ID_pointer);
-  PRECONDITION(update_in_place!=update_in_placet::MAY_UPDATE_IN_PLACE);
+  PRECONDITION(expr.type().id() == ID_pointer);
```
:pick: Avoid putting formatting changes mixed in with regular changes as it makes reviewing hard.

antlechner replied:
Different reviewers seem to have different opinions on that, I've received the exact opposite comment before. :confused: I guess from now on I'll put them in separate commits and maybe squash them before merging.

thomasspriggs replied:
Clang format should be run on the lines each commit updates and squashed into that commit. Clang formatting of unchanged lines should either not be done, or be done in a separate commit. I think this was commented on, because this commit is for functional changes and the only change to this line in this commit is for formatting.

antlechner replied:
Right, I used CLion's "clang-format selected lines" feature, and probably just included those two extra lines as I think CI has complained before about lines next to some lines that I changed. Will separate this in the future.

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -384,67 +385,95 @@ void java_object_factoryt::gen_pointer_target_init(
-    exprt target;
-    if(update_in_place==update_in_placet::NO_UPDATE_IN_PLACE)
+    const auto &target_class_type = to_java_class_type(target_type);
+    if(has_prefix(id2string(target_class_type.get_tag()), "java::array["))
```
:pick: Avoid incidental code changes in the main commit
:pick: use `is_valid_java_array(target_class_type)`

antlechner replied:
1) Sorry, it was a small logical change but I should have realised git would make this look hideous.
2) Done in a separate commit.

antlechner replied:
@thk123 PS I don't know why but `is_valid_java_array` broke something here. Changing this here causes something to get stuck when I run `ctest -V -L CORE -R jbmc` - it tells me how many tests it's about to run and then does nothing. I deleted the commit where I made this change, and ctest runs as usual, with all tests passing. Do you have any idea why this happens?

antlechner replied:
Some more info: the very first test run by ctest does not terminate after this change (`ArithmeticException1`).

thk123 replied:
Hmm OK, the comment isn't critical so lets investigate why after this PR is merged.

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -384,67 +385,95 @@ void java_object_factoryt::gen_pointer_target_init(
+      // the array that is returned by Enum.values().
+
+      const irep_idt &class_name = target_class_type.get_name();
+      const symbolt &values = ns.lookup(id2string(class_name) + ".$VALUES");
```
:no_entry_sign: This will crash if this symbol can't be found - to aid debugging check the symbol exists first

antlechner replied:
Added an invariant.

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -384,67 +385,95 @@ void java_object_factoryt::gen_pointer_target_init(
+      dereference_exprt deref_expr(values.symbol_expr());
+      const auto &deref_struct_type =
+        to_struct_type(ns.follow(deref_expr.type()));
+      const auto &comps = deref_struct_type.components();
```
:no_entry_sign: Add an invariant here the struct is what you're expecting `INVARIANT(is_valid_java_array(deref_struct_type)` (will make your operations below assuming it has a data and a length component more reasonable and more readable

antlechner replied:
Added a precondition (I don't think specifying a reason in an invariant is necessary as it's mentioned three times above that `$VALUES` is an array).

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -384,67 +385,95 @@ void java_object_factoryt::gen_pointer_target_init(
+      const auto &deref_struct_type =
+        to_struct_type(ns.follow(deref_expr.type()));
+      const auto &comps = deref_struct_type.components();
+      const member_exprt length_expr(deref_expr, "length", comps[1].type());
```
If you ask @thomasspriggs nicely - he might upstream this function: https://github.com/diffblue/test-gen/pull/2151/files/e9459fced6830dac77d2a1675791cd7813fd84f9..5a77d49dca52e520a091ca3933cf5db6f08e3fe1#diff-eee200ad05cc782a7f52ac864ed9130aR208 which would make this easier.

thomasspriggs replied:
PR is here - https://github.com/diffblue/cbmc/pull/2710

antlechner replied:
Thanks @thomasspriggs - let me know if/when it's merged and I'll use it here (possibly in a new PR).

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -384,67 +385,95 @@ void java_object_factoryt::gen_pointer_target_init(
+        "enum_index_init",
+        from_integer(0, java_int_type()),
+        length_expr,
+        false);
```
:no_entry_sign: If you do leave as a floating boolean parameter, please declare above the call:
```C++
const bool inclusive_max = false;
const symbol_exprt &index_expr = gen_nondet_int_init(
   assignments,
   "enum_index_init",
    from_integer(0, java_int_type()),
    length_expr,
    inclusive_max);
```

antlechner replied:
Got rid of the parameter.

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -384,67 +385,95 @@ void java_object_factoryt::gen_pointer_target_init(
+        length_expr,
+        false);
+
+      // Index array using pointer arithmetic
```
:bulb: re-write this comment as:
```C++
// generate statement using pointer arithmetic to access array element:
// expr = (expr.type())*(enum_array_expr + index_expr);

antlechner replied:
Done

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -384,67 +385,95 @@ void java_object_factoryt::gen_pointer_target_init(
+        false);
+
+      // Index array using pointer arithmetic
+      plus_exprt plus(enum_array_expr, index_expr, enum_array_expr.type());
```
:pick: No need to specify the type of the expression, it will be inferred from the lhs parameter.

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -384,67 +385,95 @@ void java_object_factoryt::gen_pointer_target_init(
+
+      // Index array using pointer arithmetic
+      plus_exprt plus(enum_array_expr, index_expr, enum_array_expr.type());
+      const dereference_exprt arraycellref(
```
:pick: `array_cell_ref`

antlechner replied:
This part is copied from somewhere else in the same file and I haven't had time to refactor it for this PR. I'll leave the name for now so if I don't get round to doing a follow-up refactoring PR, the two identical parts can be easily found in the future.

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -384,67 +385,95 @@ void java_object_factoryt::gen_pointer_target_init(
+      // Index array using pointer arithmetic
+      plus_exprt plus(enum_array_expr, index_expr, enum_array_expr.type());
+      const dereference_exprt arraycellref(
+        plus_exprt(enum_array_expr, index_expr, enum_array_expr.type()),
```
:no_entry_sign: Use your `plus` rather than remaking it!

antlechner replied:
Whoops, missed that while tidying things up. Fixed now.

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -384,67 +385,95 @@ void java_object_factoryt::gen_pointer_target_init(
+      plus_exprt plus(enum_array_expr, index_expr, enum_array_expr.type());
+      const dereference_exprt arraycellref(
+        plus_exprt(enum_array_expr, index_expr, enum_array_expr.type()),
+        enum_array_expr.type().subtype());
```
:pick: Don't specify the type, the single parameter constructor will do what you want and is less error prone.

thomasspriggs replied:
In the case of a reference typed array, the type may need to be explicitly specified.

thk123 replied:
Why? If it does, it suggests a type cast is needed, if the type if the `expr` passed in as the first parameter is T*, then any well formed `dereference_exprt` will have type `T` (i.e. `expr.type().subtype()`)

antlechner replied:
@thk123 I removed the type. I checked the two constructors and the single parameter version does just call the other one with `op.type().subtype()`.

thk123 replied:
Yeah the only way this could be different is if the `type()` of `plus` is different from `array_enum_expr`. The type of `plus` is determines by the lhs expr so in this case you're definitely fine. I think it would be valid to do `index_expr + enum_array_expr`  and then you would get the wrong answer...

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -384,67 +385,95 @@ void java_object_factoryt::gen_pointer_target_init(
+      return;
     }
-    else
+    if(target_class_type.get_base("java::java.lang.Enum"))
```
:bulb: Pull this out into into a `gen_nondet_enum_init`

antlechner replied:
Done in a separate commit

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -384,67 +385,95 @@ void java_object_factoryt::gen_pointer_target_init(
-      typet(), // override type immaterial
-      depth+1,
-      update_in_place);
+  // obtain a target pointer to initialize; if in MUST_UPDATE_IN_PLACE mode we
```
Note to self/reviewers - this code hasn't changed, just GH messing up the move.

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1280,7 +1315,8 @@ void java_object_factoryt::allocate_nondet_length_array(
     from_integer(0, java_int_type()),
-    max_length_expr);
+    max_length_expr,
+    true);
```
:no_entry_sign: Pull into a const variable so documenting what this floating boolean is

antlechner replied:
Got rid of the boolean.

thk123 commented on jbmc/src/java_bytecode/ci_lazy_methods.cpp:
```diff
@@ -382,6 +382,8 @@ void ci_lazy_methodst::initialize_instantiated_classes(
       if(param.type().id()==ID_pointer)
       {
         const pointer_typet &original_pointer=to_pointer_type(param.type());
+        add_clinit_call_for_pointer_type(
```
:no_entry_sign: I can't remember exactly where it is, but somewhere we try using the `pointer_type_selector`, you should probably `add_clinit_call_for_pointer_type` for the replacement pointer too.

thk123 replied:
Nope - this is fine because `add_all_needed_classes` calls `initialize_instantiated_classes` with both the original and any substituted pointer

thk123 reviewed (changes requested):
Additional tests needed for your change to always call static init:

- static field assigned a value then its value should be set at the start of the function
 - (TG test): static field on a class that implements an interface that gets substituted should be set up. 

*Emojis explained*
:no_entry_sign: - must be addressed
:question: - must be answered
:bulb: - Apply the ones you agree with and have time for
:pick: - Apply if it is easy and you are already making changes. Mostly adding so these things can be avoided in future

Reviewed commit by commit so some relevant comments may be hidden

Oh BTW seems really negative review, but this no the whole looks great and I'm looking forward to this feature!

antlechner committed:
Move nondet enum initialization to new function

antlechner commented:
@thk123 I think I implemented and/or replied to all of your comments. Except for tests - I don't think the difference between what we used to do and what we do after these changes can be observed in regression tests: as soon as we try to observe a static variable we have to access it and at that point we would have always called `clinit_wrapper` anyway. I did however add some more tests that check that not only are enums initialised to be some constant of their type, but in fact any of these constants can be chosen. I also rebased on develop, so this is ready for re-review.

thk123 reviewed (approved):
Looks great, thanks for all the improvements - needs a TG bump if it hasn't got one already

thk123 commented:
Has one: diffblue/test-gen#2156 - currently failing CI

peterschrammel commented on jbmc/regression/jbmc/NondetEnumArg/NondetEnumArg.java:
```diff
@@ -0,0 +1,46 @@
+        if (c == null)
+            return;
+        assert c != null;
+        boolean isRed = c.name().startsWith("RED") && c.name().length() == 3
```
May we have tests with `switch` already?

antlechner replied:
We have one tests for `switch` on an enum type. It uses `--cover location` and pretty much just checks that all possible ordinals (except the last one) can be generated for a switch statement.

antlechner replied:
@peterschrammel Do you think the `switch` test and the new `NondetEnumArg` test should be combined?

peterschrammel reviewed (approved):
LGTM modulo TG passing

antlechner committed:
Temporarily remove new UNREACHABLE statements

These really should be unreachable, but some company code needs to be
changed for this first. A ticket number has been added as a reminder.

antlechner committed:
Restrict new clinit_wrapper calls to enum types

Calling clinit_wrapper on all nondet-initialized types has a noticeable
impact on performance, so for now we are only doing this for enums, and
we will re-evaluate the behavior for other types in the future.

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -785,18 +785,28 @@ void java_object_factoryt::gen_nondet_pointer_init(
-  if(!must_be_null)
+  // case in gen_nondet_struct_init would slow symex down too much, so if we
+  // decide to do this for all types, we should do it here.
+  if(can_cast_type<java_class_typet>(subtype))
```
:pick: mild preference for:
```
if(const auto class_type = type_try_dynamic_cast<java_class_typet>(subtype)
{
  if(class_type->get_base("java::java.lang.Enum") && !must_be_null)
```

Since puts in one place the check that is is a `java_class_typet`

thk123 commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -785,18 +785,28 @@ void java_object_factoryt::gen_nondet_pointer_init(
-  // case in gen_nondet_struct_init would slow symex down too much.
-  if(!must_be_null)
+  // case in gen_nondet_struct_init would slow symex down too much, so if we
+  // decide to do this for all types, we should do it here.
```
Add a comment saying that this is mirrored in `ci_lazy_methods`

thk123 reviewed (approved):
My only issue is we are duplicating the logic for when we call `clinit`. I can't immediately think of a better way (and you have a comment linking the two. You could pull out a weird utility function like `does_need_clinit_call` so they both use the same condition.

antlechner commented:
@thk123 Applied your two line-specific comments. The duplication is a general problem with CI lazy methods, which has to mirror everything that happens in `__CPROVER_initialize` and `__CPROVER_start`. Symex-driven lazy loading no longer has this problem. I'll check if it's worth adding a shared utility function for the "is this an enum?" logic in a follow-up PR.
